board/batocera/rockchip/rk3326/linux_patches/linux-004-rg351-controller-rumble.patch

```diff
@@ -0,0 +1,290 @@
++			fwnode_property_read_u32(child, "abs-fuzz", &axes[i].fuzz);
++			fwnode_property_read_u32(child, "abs-flat", &axes[i].flat);
++			input_set_abs_params(joy->input, axes[i].code,
++					axes[i].range[0], axes[i].range[1],
++					axes[i].fuzz, axes[i].flat);
++			input_set_capability(joy->input, EV_ABS, axes[i].code);
++		}
++		else
++		{
++			axes[i].is_analog = 0;
++			axes[i].gpiod = devm_fwnode_gpiod_get(dev, child, NULL, GPIOD_IN, "btn");
++			input_set_capability(joy->input, EV_KEY, axes[i].code);
++		}
++
++	}
++
++	joy->axes = axes;
++	return 0;
++}
++
++static void adc_joystick_polled_poll(struct input_dev *input)
++{
++	struct adc_joystick_polled *bdev = input_get_drvdata(input);
++	int i;
++
++	int raw;
++
++	for (i = 0; i < bdev->num_chans; i++) {
++		if (bdev->axes[i].is_analog == 1)
++		{
++			iio_read_channel_raw(&bdev->chans[i], &raw);
++			input_event(input, EV_ABS, bdev->axes[i].code, raw);
++		}
++		else
++		{
++			raw = gpiod_get_value_cansleep(bdev->axes[i].gpiod);
++			if (raw != bdev->axes[i].prev_val)
++			{
++				input_event(input, EV_KEY, bdev->axes[i].code, raw);
++				bdev->axes[i].prev_val = raw;
++			}
++		}
++	}
++
++	input_sync(input);
++}
++
++static int adc_joystick_polled_probe(struct platform_device *pdev)
++{
++	struct device *dev = &pdev->dev;
++	struct adc_joystick_polled *joy;
++	struct input_dev *input;
++	u32 joypad_revision = 0;
++	u32 joypad_product = 0;
++	int err;
++
++	joy = devm_kzalloc(dev, sizeof(*joy), GFP_KERNEL);
++	if (!joy)
++		return -ENOMEM;
++
++	joy->chans = devm_iio_channel_get_all(dev);
++	if (IS_ERR(joy->chans)) {
++		err = PTR_ERR(joy->chans);
++		if (err != -EPROBE_DEFER)
++			dev_err(dev, "Unable to get IIO channels");
++		return err;
++	}
++
++	input = devm_input_allocate_device(dev);
++	if (!input)
++		return -ENOMEM;
++
++	joy->input = input;
++	device_property_read_string(dev, "joypad-name", &input->name);
++	device_property_read_u32(dev, "joypad-revision", &joypad_revision);
++	device_property_read_u32(dev, "joypad-product", &joypad_product);
++	input->id.bustype = BUS_HOST;
++	input->id.product = (u16)joypad_product;
++	input->id.version = (u16)joypad_revision;
++
++	input_setup_polling(input, adc_joystick_polled_poll);
++	input_set_poll_interval(input, 10);
++	
++	adc_joystick_set_axes(dev, joy);	
++
++	adc_joystick_set_rumble(dev, joy);
++
++	input_set_drvdata(input, joy);
++
++	input_register_device(input);
++
++	dev_info(dev, "adc-joystick-polled: Bound to device correctly!");
++	return 0;
++}
++
++static const struct of_device_id adc_joystick_polled_of_match[] = {
++	{ .compatible = "gpio-keys-adc", },
++	{ }
++};
++
++static struct platform_driver adc_joystick_polled_driver = {
++	.driver = {
++		.name = "adc-joystick-polled",
++		.of_match_table = adc_joystick_polled_of_match,
++	},
++	.probe = adc_joystick_polled_probe,
++};
++module_platform_driver(adc_joystick_polled_driver);
++
++MODULE_AUTHOR("Tony Jih <tonyjih@gmail.com>");
```

Comment:
Technically I authored this driver but what i left it as was about 150 lines long.
You call if you're changing that MODULE_AUTHOR string.
Just remember that what you do now will be remembered in future :)